File: packages/core/src/zip.ts
import type { ObjectPatchDiff } from './diff'
import type { ObjectStream } from './stream'

/**
 * A compressed representation of a patch operation.
 */
export type CompressedObjectPatchDiff = [1 | 3, number | string, any] | [2, number | string]

/**
 * A compressed message in the object stream.
 */
export type CompressedObjectStream<T extends object, E> =
  /** Initial state. */
  | [1, T]
  /** Compressed delta. */
  | [2, CompressedObjectPatchDiff]
  /** Final state. */
  | [3, T]
  /** Event. */
  | [9, E]

/**
 * Compresses a stream of object messages.
 *
 * @param messages - The stream of messages to compress.
 * @returns An async iterable yielding compressed messages.
 */
export async function* deflate<T extends object, E>(
  messages: AsyncIterable<ObjectStream<T, E>>
): AsyncIterable<CompressedObjectStream<T, E>> {
  const objType = { init: 1, delta: 2, done: 3, event: 9 } as const
  const patchType = { a: 1, d: 2, s: 3 } as const // Add, Remove, Replace RFC6902
  const deltaDict: Record<string, number> = {}
  let deltaOffset = 0

  for await (const msg of messages) {
    if (msg.t === 'delta') {
      // Delta compression
      const delta: any = []
      for (const [typ, path, ...rest] of msg.d) {
        delta.push([
          patchType[typ],
          // Previous Occurrence Compression
          path in deltaDict ? deltaOffset - deltaDict[path] : path,
          ...rest
        ])
        deltaDict[path] = deltaOffset
        deltaOffset += 1
      }
      yield [objType.delta, delta]
    } else {
      yield (msg.d ? [objType[msg.t], msg.d] : [objType[msg.t]]) as any
    }
  }
}

/**
 * Decompresses a stream of compressed object messages.
 *
 * @param compressed - The stream of compressed messages to decompress.
 * @yields Decompressed object stream messages.
 */
export async function* inflate<T extends object, E>(
  compressed: AsyncIterable<CompressedObjectStream<T, E>>
): AsyncIterable<ObjectStream<T, E>> {
  const objType = { 1: 'init', 2: 'delta', 3: 'done', 9: 'event' } as any
  const patchType = { 1: 'a', 2: 'd', 3: 's' } as any // Add, Remove, Replace RFC6902
  const deltaDict: Record<number, string> = {}
  const deltaDictInv: Record<string, number> = {}
  let deltaOffset = 0

  for await (const rec of compressed) {
    // Skip compression if the payload does not seem a compressed one. This will
    // simplify the client and make it retro compatible
    if (!Array.isArray(rec)) {
      yield rec
      continue
    }
    const [typ, data] = rec
    if (objType[typ] === 'delta') {
      const delta: ObjectPatchDiff = []
      for (const [typ, path, ...rest] of data as CompressedObjectPatchDiff) {
        let decPath = path
        if (typeof path === 'number') {
          const oldOffset = deltaOffset - path
          decPath = deltaDict[oldOffset]
          delete deltaDict[oldOffset]
        }
        // This is needed to kep the main dictionary void of any duplicates. This may
        // be dangerous in case the client is not sending a compressed stream.
        if (decPath in deltaDictInv) delete deltaDict[deltaDictInv[decPath]]
        deltaDictInv[decPath] = deltaOffset
        //
        deltaDict[deltaOffset] = decPath
        deltaOffset += 1

        delta.push([patchType[typ], decPath, ...rest] as any)
      }
      yield { t: 'delta', d: delta }
    } else {
      yield { t: objType[typ], d: data as any }
    }
  }
}

File: python/diffcp/zip.py
from collections.abc import AsyncIterable
from typing import Any, TypeVar

T = TypeVar('T')
E = TypeVar('E')

# CompressedObjectPatchDiff = [1 | 3, number | string, any] | [2, number | string]
CompressedObjectPatchDiff = tuple[int, int | str, Any] | tuple[int, int | str]

# CompressedObjectStream = [1, T] | [2, List[CompressedObjectPatchDiff]] | [3, T] | [9, E]
CompressedObjectStream = tuple[int, Any] | tuple[int]

OBJ_TYPE_MAP = {'init': 1, 'delta': 2, 'done': 3, 'event': 9}
OBJ_TYPE_INV = {v: k for k, v in OBJ_TYPE_MAP.items()}
PATCH_TYPE_MAP = {'a': 1, 'd': 2, 's': 3}
PATCH_TYPE_INV = {v: k for k, v in PATCH_TYPE_MAP.items()}


async def deflate(messages: AsyncIterable[dict[str, Any]]) -> AsyncIterable[CompressedObjectStream]:
    delta_dict: dict[str, int] = {}
    delta_offset = 0

    async for msg in messages:
        t = msg['t']
        d = msg.get('d')

        if t == 'delta':
            delta_payload = []
            for op in d:
                typ = op[0]
                path = op[1]
                rest = op[2:]

                # Previous Occurrence Compression
                if path in delta_dict:
                    path_val: int | str = delta_offset - delta_dict[path]
                else:
                    path_val = path

                delta_payload.append([PATCH_TYPE_MAP[typ], path_val, *rest])
                delta_dict[path] = delta_offset
                delta_offset += 1
            yield [OBJ_TYPE_MAP['delta'], delta_payload]
        else:
            if d is not None:
                yield [OBJ_TYPE_MAP[t], d]
            else:
                yield [OBJ_TYPE_MAP[t]]


async def inflate(
    compressed: AsyncIterable[dict[str, Any] | CompressedObjectStream],
) -> AsyncIterable[dict[str, Any]]:
    delta_dict: dict[int, str] = {}
    delta_dict_inv: dict[str, int] = {}
    delta_offset = 0

    async for rec in compressed:
        # Skip compression if the payload does not seem a compressed one (e.g. if it's already a dict)
        if isinstance(rec, dict):
            yield rec
            continue

        typ_code = rec[0]
        data = rec[1] if len(rec) > 1 else None

        t = OBJ_TYPE_INV.get(typ_code)
        if t == 'delta':
            delta_ops = []
            for op in data:
                op_typ_code = op[0]
                path_val = op[1]
                rest = op[2:]

                dec_path = path_val
                if isinstance(path_val, int):
                    old_offset = delta_offset - path_val
                    dec_path = delta_dict[old_offset]
                    del delta_dict[old_offset]

                # Cleanup duplicates
                if dec_path in delta_dict_inv:
                    inv_offset = delta_dict_inv[dec_path]
                    if inv_offset in delta_dict:
                        del delta_dict[inv_offset]

                delta_dict_inv[dec_path] = delta_offset
                delta_dict[delta_offset] = dec_path
                delta_offset += 1

                op_typ = PATCH_TYPE_INV[op_typ_code]
                delta_ops.append([op_typ, dec_path, *rest])
            yield {'t': 'delta', 'd': delta_ops}
        else:
            msg = {'t': t}
            if data is not None:
                msg['d'] = data
            yield msg
